src/memray/commands/attach.py

```diff
@@ -12,22 +12,49 @@
 import subprocess
 import sys
 import threading
+from enum import IntEnum
 
 import memray
 from memray._errors import MemrayCommandError
 
 from .live import LiveCommand
 from .run import _get_free_port
 
+
+class TrackingMode(IntEnum):
+    ACTIVATE = 1
+    DEACTIVATE = 2
+    UNTIL_HEAP_SIZE = 3
+    FOR_DURATION = 4
+
+
 GDB_SCRIPT = pathlib.Path(__file__).parent / "_attach.gdb"
 LLDB_SCRIPT = pathlib.Path(__file__).parent / "_attach.lldb"
 RTLD_DEFAULT = memray._memray.RTLD_DEFAULT
 RTLD_NOW = memray._memray.RTLD_NOW
 PAYLOAD = """
 import atexit
+import time
+import threading
+import resource
+import logging
+import pathlib
+from typing import Set
+from typing import Union
+from contextlib import suppress
 
 import memray
+from memray.commands.attach import TrackingMode
+
+ACTIVE_THREADS: Set[Union[threading.Thread, threading.Timer]] = set()
+TRACKER_LOCK = threading.Lock()
```

Comment:
This isn't going to work the way you want. You're trying to share these globals across multiple calls to attach, but each attach call runs its own script, with its own globals namespace. That's why we do the `memray._last_tracker = ...` thing below: we're sticking the tracker, which we need to be shared across calls to `memray attach`, in a global variable of the `memray` module, so that future calls from different namespaces can share it.

Honestly, this is probably a reason to split this out into a proper module. This is getting too complicated to manage as an inline script like this... I'm going to take a swing at refactoring this and giving you something easier to work with for basing your changes on top of...

Reply:
OK, the last commit on https://github.com/bloomberg/memray/pull/459 gives you something saner to base this on. Rather than just a giant string literal, that PR will give you a new module named `memray._attach_callback` that we can put these changes in, to help make managing the global state easier.

The `memray._attach_callback.callback` function will be the entry point. It'll take both the arguments that are going to be passed to the `Tracker` constructor as well as the arguments determining which attach mode to use, and it will call other functions in the module to handle the tracker creation and destruction, as well as the threads to disable tracking when different thresholds are reached.